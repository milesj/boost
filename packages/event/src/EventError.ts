import { createScopedError } from '@boost/internal';

const errors = {
  LISTENER_INVALID: 'Invalid event listener for "%s", must be a function.',
  NAME_INVALID:
    'Invalid event %s "%s". May only contain dashes, periods, and lowercase characters.',
};

export type EventErrorCode = keyof typeof errors;

export default createScopedError<EventErrorCode>('EVT', 'EventError', errors);
