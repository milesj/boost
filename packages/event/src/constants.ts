export const WILDCARD_SCOPE = '*';

export const EVENT_NAME_PATTERN = /^[a-z]{1}[-.a-z0-9]*[a-z]{1}$/u;
