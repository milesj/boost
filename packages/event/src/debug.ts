import { createInternalDebugger } from '@boost/internal';

export default createInternalDebugger('event');
