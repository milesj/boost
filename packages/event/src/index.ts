/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './BailEvent';
export * from './BaseEvent';
export * from './ConcurrentEvent';
export * from './constants';
export * from './Event';
export * from './EventError';
export * from './types';
export * from './WaterfallEvent';
