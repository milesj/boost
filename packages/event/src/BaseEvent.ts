import EventError from './EventError';
import debug from './debug';
import { EVENT_NAME_PATTERN, WILDCARD_SCOPE } from './constants';
import { Listener, WildstarScope } from './types';

export default abstract class BaseEvent<
  Return,
  Args extends unknown[],
  Scope extends string = string
> {
  listeners = new Map<Scope | WildstarScope, Set<Listener<Args, Return>>>();

  name: string;

  constructor(name: string) {
    this.name = this.validateName(name, 'name');

    debug('New %S created: %s', this.constructor.name, name);
  }

  /**
   * Remove all listeners from the event.
   */
  clearListeners(scope?: Scope): this {
    if (scope) {
      this.getListeners(scope).clear();
    } else {
      this.listeners.clear();
    }

    return this;
  }

  /**
   * Return a set of listeners for a specific event scope.
   */
  getListeners(scope?: Scope): Set<Listener<Args, Return>> {
    const key = this.validateName(scope || WILDCARD_SCOPE, 'scope');

    if (!this.listeners.has(key)) {
      this.listeners.set(key, new Set());
    }

    return this.listeners.get(key)!;
  }

  /**
   * Return a list of all scopes with listeners.
   */
  getScopes(): (Scope | WildstarScope)[] {
    return Array.from(this.listeners.keys()) as (Scope | WildstarScope)[];
  }

  /**
   * Register a listener to the event.
   */
  listen(listener: Listener<Args, Return>, scope?: Scope): this {
    debug('Registering "%s" listener', this.name);

    this.getListeners(scope).add(this.validateListener(listener));

    return this;
  }

  /**
   * Register a listener to the event that only triggers once.
   */
  once(listener: Listener<Args, Return>, scope?: Scope): this {
    const func = this.validateListener(listener);
    const handler = ((...args: unknown[]) => {
      this.unlisten(handler);

      return func(...args);
    }) as Listener<Args, Return>;

    return this.listen(handler, scope);
  }

  /**
   * Remove a listener from the event.
   */
  unlisten(listener: Listener<Args, Return>, scope?: Scope): this {
    debug('Unregistering "%s" listener', this.name);

    this.getListeners(scope).delete(listener);

    return this;
  }

  /**
   * Validate the listener is a function.
   */
  protected validateListener<L>(listener: L): L {
    if (typeof listener !== 'function') {
      throw new EventError('LISTENER_INVALID', [this.name]);
    }

    return listener;
  }

  /**
   * Validate the name/scope match a defined pattern.
   */
  protected validateName<N extends string>(name: N, type: string): N {
    if (type === 'scope' && name === WILDCARD_SCOPE) {
      return name;
    }

    if (!name.match(EVENT_NAME_PATTERN)) {
      throw new EventError('NAME_INVALID', [type, name]);
    }

    return name;
  }

  /**
   * Emit the event by executing all scoped listeners with the defined arguments.
   */
  abstract emit(args: unknown, scope?: Scope): unknown;
}
