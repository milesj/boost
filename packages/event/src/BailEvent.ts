import BaseEvent from './BaseEvent';
import debug from './debug';

export default class BailEvent<
  Args extends unknown[],
  Scope extends string = string
> extends BaseEvent<boolean | void, Args, Scope> {
  /**
   * Synchronously execute listeners with the defined arguments.
   * If a listener returns `false`, the loop with be aborted early,
   * and the emitter will return `true` (for bailed).
   */
  emit(args: Args, scope?: Scope): boolean {
    if (__DEV__) {
      debug('Emitting "%s%s" as bail', this.name, scope ? `:${scope}` : '');
    }

    return Array.from(this.getListeners(scope)).some((listener) => listener(...args) === false);
  }
}
