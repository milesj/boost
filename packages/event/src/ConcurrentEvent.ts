import BaseEvent from './BaseEvent';
import debug from './debug';

export default class ConcurrentEvent<
  Args extends unknown[],
  Scope extends string = string
> extends BaseEvent<Promise<unknown>, Args, Scope> {
  /**
   * Asynchronously execute listeners for with the defined arguments.
   * Will return a promise with an array of each listener result.
   */
  emit(args: Args, scope?: Scope): Promise<unknown[]> {
    debug('Emitting "%s%s" as concurrent', this.name, scope ? `:${scope}` : '');

    return Promise.all(Array.from(this.getListeners(scope)).map((listener) => listener(...args)));
  }
}
