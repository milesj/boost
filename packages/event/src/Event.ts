import { BaseEvent } from './BaseEvent';

export class Event<Args extends unknown[], Scope extends string = string> extends BaseEvent<
	void,
	Args,
	Scope
> {
	/**
	 * Synchronously execute listeners with the defined arguments.
	 */
	emit(args: Args, scope?: Scope) {
		[...this.getListeners(scope)].forEach((listener) => {
			listener(...args);
		});
	}
}
