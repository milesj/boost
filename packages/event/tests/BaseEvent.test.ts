import { beforeEach, describe, expect, it } from 'vitest';
import { Event } from '../src/Event';

describe('Event', () => {
	let event: Event<unknown[]>;

	beforeEach(() => {
		event = new Event('event.test');
	});

	describe('constructor()', () => {
		it('errors if name contains invalid characters', () => {
			expect(() => new Event('foo+bar')).toThrowErrorMatchingSnapshot();
		});
	});

	describe('clearListeners()', () => {
		it('deletes all listeners in a scope', () => {
			event.listen(() => {}, 'foo');
			event.listen(() => {}, 'bar');
			event.listen(() => {}, 'baz');

			expect(event.getListeners('foo').size).toBe(1);
			expect(event.getListeners('bar').size).toBe(1);
			expect(event.getListeners('baz').size).toBe(1);

			event.clearListeners('foo');

			expect(event.getListeners('foo').size).toBe(0);
			expect(event.getListeners('bar').size).toBe(1);
			expect(event.getListeners('baz').size).toBe(1);
		});

		it('deletes all listeners across all scopes', () => {
			event.listen(() => {}, 'foo');
			event.listen(() => {}, 'bar');
			event.listen(() => {}, 'baz');

			expect(event.getListeners('foo').size).toBe(1);
			expect(event.getListeners('bar').size).toBe(1);
			expect(event.getListeners('baz').size).toBe(1);

			event.clearListeners();

			expect(event.getListeners('foo').size).toBe(0);
			expect(event.getListeners('bar').size).toBe(0);
			expect(event.getListeners('baz').size).toBe(0);
		});
	});

	describe('getListeners()', () => {
		it('errors if scope contains invalid characters', () => {
			expect(() => event.getListeners('foo+bar')).toThrowErrorMatchingSnapshot();
		});

		it('doesnt error for default scope', () => {
			expect(() => event.getListeners()).not.toThrow();
		});

		it('creates the listeners set if it does not exist', () => {
			expect(event.listeners.has('foo')).toBe(false);

			const set = event.getListeners('foo');

			expect(set).toBeInstanceOf(Set);
			expect(event.listeners.has('foo')).toBe(true);
		});
	});

	describe('getScopes()', () => {
		it('returns an array of scope names', () => {
			event.getListeners();
			event.getListeners('foo');
			event.getListeners('bar');
			event.getListeners('baz');

			expect(event.getScopes()).toEqual(['*', 'foo', 'bar', 'baz']);
		});
	});

	describe('listen()', () => {
		it('errors if listener is not a function', () => {
			expect(() => {
				// @ts-expect-error Invalid type
				event.listen(123);
			}).toThrowErrorMatchingSnapshot();
		});

		it('adds a listener to the default scope', () => {
			const listener = () => {};

			expect(event.getListeners().has(listener)).toBe(false);

			event.listen(listener);

			expect(event.getListeners().has(listener)).toBe(true);
		});

		it('adds a listener to a specific scope', () => {
			const listener = () => {};

			expect(event.getListeners('foo').has(listener)).toBe(false);

			event.listen(listener, 'foo');

			expect(event.getListeners('foo').has(listener)).toBe(true);
		});

		it('removes a listener from the default scope', () => {
			const listener = () => {};

			const unlisten = event.listen(listener);

			expect(event.getListeners().has(listener)).toBe(true);

			unlisten();

			expect(event.getListeners().has(listener)).toBe(false);
		});

		it('removes a listener from a specific scope', () => {
			const listener = () => {};

			const unlisten = event.listen(listener, 'foo');

			expect(event.getListeners('foo').has(listener)).toBe(true);

			unlisten();

			expect(event.getListeners('foo').has(listener)).toBe(false);
		});
	});

	describe('unlisten()', () => {
		it('removes a listener from the default scope', () => {
			const listener = () => {};

			event.listen(listener);

			expect(event.getListeners().has(listener)).toBe(true);

			event.unlisten(listener);

			expect(event.getListeners().has(listener)).toBe(false);
		});

		it('removes a listener from a specific scope', () => {
			const listener = () => {};

			event.listen(listener, 'foo');

			expect(event.getListeners('foo').has(listener)).toBe(true);

			event.unlisten(listener, 'foo');

			expect(event.getListeners('foo').has(listener)).toBe(false);
		});
	});

	describe('once()', () => {
		it('errors if listener is not a function', () => {
			expect(() => {
				// @ts-expect-error Invalid type
				event.once(123);
			}).toThrowErrorMatchingSnapshot();
		});

		it('adds a listener to the default scope', () => {
			const listener = () => {};

			expect(event.getListeners().size).toBe(0);

			event.once(listener);

			expect(event.getListeners().has(listener)).toBe(false);
			expect(event.getListeners().size).toBe(1);
		});

		it('adds a listener to a specific scope', () => {
			const listener = () => {};

			expect(event.getListeners('foo').size).toBe(0);

			event.once(listener, 'foo');

			expect(event.getListeners('foo').has(listener)).toBe(false);
			expect(event.getListeners('foo').size).toBe(1);
		});

		it('removes the listener once executed', () => {
			let count = 0;
			const listener = () => {
				count += 1;
			};

			event.once(listener);

			expect(event.getListeners().size).toBe(1);

			event.emit([]);
			event.emit([]);
			event.emit([]);

			expect(event.getListeners().size).toBe(0);
			expect(count).toBe(1);
		});
	});
});
