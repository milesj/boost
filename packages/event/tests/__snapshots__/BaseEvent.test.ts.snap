// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event constructor() errors if name contains invalid characters 1`] = `"Invalid event name \\"foo+bar\\". May only contain dashes, periods, and lowercase characters. [EVT:NAME_INVALID]"`;

exports[`Event getListeners() errors if scope contains invalid characters 1`] = `"Invalid event scope \\"foo+bar\\". May only contain dashes, periods, and lowercase characters. [EVT:NAME_INVALID]"`;

exports[`Event listen() errors if listener is not a function 1`] = `"Invalid event listener for \\"event.test\\", must be a function. [EVT:LISTENER_INVALID]"`;

exports[`Event once() errors if listener is not a function 1`] = `"Invalid event listener for \\"event.test\\", must be a function. [EVT:LISTENER_INVALID]"`;
